Fix flaky TestApplyCRDuringCRDFinalization test

aviralgarg05 opened the pull request:
#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature


Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
--> /kind flake

#### What this PR does / why we need it:
This PR fixes the flaky TestApplyCRDuringCRDFinalization test that was failing 
intermittently on slower systems (s390x architecture, race detector builds).

The root cause was a race condition where the test would attempt to apply a CR
immediately after requesting CRD deletion, without waiting for the CRD to
actually enter the terminating state. The fix explicitly waits for the CRD to
have a DeletionTimestamp set before attempting the apply operation.
#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
--> #135403 

#### Special notes for your reviewer:

The test was run multiple times (5x without race detector, 3x with race detector)
to verify the fix eliminates the flakiness. All related finalization tests
(TestFinalization, TestFinalizationAndDeletion, TestApplyCRDuringCRDFinalization)
pass consistently.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".


For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
--> 
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.: NA

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

pohly commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go:
```diff
@@ -179,20 +179,25 @@ func TestApplyCRDuringCRDFinalization(t *testing.T) {
 
+	// Wait for the CRD to have a DeletionTimestamp, confirming it's in terminating state.
+	// This avoids flakiness on slower systems where the deletion might not be processed
+	// before we attempt the apply.
```
This sounds more convincing to me now. But I am not familiar enough with CRDs to review this, I'd like to defer to someone else.

Rajalakshmi-Girish replied:
@pohly can you suggest someone who can review this PR?
We see the https://github.com/kubernetes/kubernetes/issues/135403 issue occuring often. [Triage Link](https://storage.googleapis.com/k8s-triage/index.html?text=TestApplyCRDuringCRDFinalization&test=k8s.io%2Fapiextensions-apiserver%2Ftest%2Fintegration&xjob=s3) shows 11 failures in last 15 days.

@aviralgarg05 ^^

aviralgarg05 commented:
Please review this pr

aviralgarg05 committed:
Fix flaky TestApplyCRDuringCRDFinalization test

This PR fixes the flaky TestApplyCRDuringCRDFinalization test that was failing
intermittently on slower systems (s390x architecture, race detector builds).

The root cause was a race condition where the test would attempt to apply a CR
immediately after requesting CRD deletion, without waiting for the CRD to
actually enter the terminating state. The fix explicitly waits for the CRD
to have the Terminating condition set to True before attempting the apply.

pohly commented:
/lgtm

Based on https://github.com/kubernetes/kubernetes/pull/135567#issuecomment-3745611276.

/test pull-kubernetes-e2e-gce

Maybe https://github.com/kubernetes/kubernetes/issues/135713

/test pull-kubernetes-verify

Test pod scheduling issue.
